crossdock/behavior/trace/thrift.go

```
@@ -0,0 +1,113 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package trace
+
+import (
+	"encoding/json"
+
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/thrift"
+	gen "github.com/uber/tchannel-go/thrift/gen-go/test"
+
+	"golang.org/x/net/context"
+)
+
+func (b *Behavior) registerThriftServer(ch *tchannel.Channel) {
+	handler := &thriftHandler{b: b, ch: ch}
+	server := thrift.NewServer(ch)
+	server.Register(gen.NewTChanSimpleServiceServer(handler))
+	b.thriftCall = handler.callDownstream
+}
+
+type thriftHandler struct {
+	ch *tchannel.Channel
+	b  *Behavior
+}
+
+func (h *thriftHandler) Call(ctx thrift.Context, arg *gen.Data) (*gen.Data, error) {
+	req, err := requestFromThrift(arg)
+	if err != nil {
+		return nil, err
+	}
+	res, err := h.b.prepareResponse(ctx, req.Downstream)
+	if err != nil {
+		return nil, err
+	}
+	return responseToThrift(res)
+}
+
+func (h *thriftHandler) Simple(ctx thrift.Context) error {
+	panic("Simple not implemented")
+}
+
+func (h *thriftHandler) callDownstream(ctx context.Context, target *Downstream) (*Response, error) {
+	req := &Request{
+		ServerRole: target.ServerRole,
+		Downstream: target.Downstream,
+	}
+	opts := &thrift.ClientOptions{HostPort: target.HostPort}
+	thriftClient := thrift.NewClient(h.ch, target.ServiceName, opts)
+	serviceClient := gen.NewTChanSimpleServiceClient(thriftClient)
+	tReq, err := requestToThrift(req)
+	if err != nil {
+		return nil, err
+	}
+
+	tctx := thrift.Wrap(ctx)
+	res, err := serviceClient.Call(tctx, tReq)
+	if err != nil {
+		panic(err.Error())
```

why panic instead of just returning the error?